### ISSUE TITLE
Suggestion: Add compiler option to disallow assignment between any and non-any types except with assertion

### ISSUE THREAD

**Issue by dsherret** [Microsoft/TypeScript#4524]
It might be useful to add a compiler option that doesn't allow assignments between any and non-any types except with an assertion.

For example, with the compiler option turned on:

```typescript
var myAny: any;
var myString: string;

myString = myAny;           // error
myString = <string> myAny;  // ok
myAny    = myString;        // error
myAny    = <any> myString;  // ok
```

It might not be worth the effort, but [sometimes people want a feature like this](http://stackoverflow.com/q/32227277/188246). Personally, I try to avoid using the `any` type in the first place except when necessary, but I would like to hear all of your thoughts on this because there might be a good case for it.

Also, sorry if this idea was raised before... I couldn't find it.

**Comment by RyanCavanaugh** [comment]
The whole point of `any` is that you don't need to do that kind of thing. It'd be very weird to have a built-in flag that completely inverted that. The way the type system works, it would be infeasible to turn this off at a global level (there are some non-visible `any` types that occur in normal use that would cause this to be an error).

It'd be reasonably straightforward to write a TSLint rule that enforced this in a more targeted way.

**Comment by dsherret** [comment]
@RyanCavanaugh yes, I agree. I actually thought of writing a tslint rule and have been looking into it briefly.

**Comment by DavidKDeutsch** [comment]
@RyanCavanaugh - would it be possible to implement this on a per-variable/argument basis? I would love to be able to do something like:

'''ts
function MyFunc(@noAny myArg: string) {
}

let x: any = "something"; 
MyFunc(any); // error
```

**Comment by RyanCavanaugh** [comment]
We're not adding anything to the language to support that sort of thing. It should be possible to write a TSLint rule that would catch that, though.

**Comment by DavidKDeutsch** [comment]
Unfortunately, TSLint can't handle such a case until [this issue](https://github.com/palantir/tslint/issues/680) is resolved; until then it can only handle syntax. I'll probably make my own custom "linter" though, that accomplishes this.
